gateway/reverse_proxy.go

```diff
@@ -1342,9 +1342,17 @@ func (p *ReverseProxy) WrappedServeHTTP(rw http.ResponseWriter, req *http.Reques
 	}
 
 	if createTransport {
+		oldTransport := p.TykAPISpec.HTTPTransport.transport
```

Comment:
Should we not be checking if `p.TykAPISpec.HTTPTransport == nil` as may be first time transport is created?

This unit test fails for me `TestReverseProxyDnsCache` with nil pointer exception, but if I change to;

```golang
if createTransport {
		var oldTransport *http.Transport

		if p.TykAPISpec.HTTPTransport != nil {
			oldTransport = p.TykAPISpec.HTTPTransport.transport
			// Prevent new idle connections to be generated.
			oldTransport.DisableKeepAlives = true
		}

		_, timeout := p.CheckHardTimeoutEnforced(p.TykAPISpec, req)
		p.TykAPISpec.HTTPTransport = p.httpTransport(timeout, rw, req, outreq)
		p.TykAPISpec.HTTPTransportCreated = time.Now()

		if oldTransport != nil {
			oldTransport.CloseIdleConnections()
		}
	}
```

it passes

Reply:
You are probably correct, for some reason our CI does not fail here 🤔 
That's a valid check will add it.

Reply:
Ok great, will give it a try.

Reply:
I tested with below tyk.conf

```json
{
    "hostname": "",
    "listen_address": "",
    "listen_port": 5000,
    "control_api_hostname": "",
    "control_api_port": 0,
    "secret": "12345",
    "node_secret": "12345",
    "pid_file_location": "",
    "allow_insecure_configs": false,
    "public_key_path": "",
    "allow_remote_config": false,
    "security": {
        "private_certificate_encoding_secret": "",
        "control_api_use_mutual_tls": false,
        "pinned_public_keys": null,
        "certificates": {
            "apis": null,
            "upstream": null,
            "control_api": null,
            "dashboard_api": null,
            "mdcb_api": null
        }
    },
    "http_server_options": {
        "skip_client_ca_announcement" : false,
        "override_defaults": false,
        "read_timeout": 0,
        "write_timeout": 0,
        "use_ssl_le": false,
        "enable_http2": false,
        "ssl_insecure_skip_verify": false,
        "enable_websockets": false,
        "server_name": "localhost",
        "min_version": 0,
        "flush_interval": 0,
        "skip_url_cleaning": false,
        "skip_target_path_escaping": false,
        "ssl_ciphers": null,
        "use_ssl": true,
        "certificates": [
            {
                "domain_name": "localhost",
                "cert_file": "/home/dom/cert.pem",
                "key_file": "/home/dom/key.pem"
            }
        ]
    },
    "reload_wait_time": 0,
    "version_header": "",
    "optimisations_use_async_session_write": false,
    "suppress_redis_signal_reload": false,
    "hash_keys": false,
    "hash_key_function": "",
    "enable_hashed_keys_listing": false,
    "min_token_length": 0,
    "enable_api_segregation": false,
    "template_path": "/home/dom/go/src/github.com/TykTechnologies/tyk/templates",
    "policies": {
        "policy_source": "",
        "policy_connection_string": "",
        "policy_record_name": "",
        "allow_explicit_policy_id": false
    },
    "disable_ports_whitelist": false,
    "ports_whitelist": null,
    "app_path": "apps/",
    "use_db_app_configs": true,
    "db_app_conf_options": {
        "connection_string": "http://localhost:3000",
        "node_is_segmented": false,
        "tags": null
    },
    "storage": {
        "type": "redis",
        "host": "localhost",
        "port": 6379,
        "hosts": null,
        "addrs": null,
        "master_name": "",
        "username": "",
        "password": "",
        "database": 0,
        "optimisation_max_idle": 100,
        "optimisation_max_active": 0,
        "timeout": 0,
        "enable_cluster": false,
        "use_ssl": false,
        "ssl_insecure_skip_verify": false
    },
    "disable_dashboard_zeroconf": false,
    "slave_options": {
        "use_rpc": false,
        "use_ssl": false,
        "ssl_insecure_skip_verify": false,
        "connection_string": "",
        "rpc_key": "",
        "api_key": "",
        "enable_rpc_cache": false,
        "bind_to_slugs": false,
        "disable_keyspace_sync": false,
        "group_id": "",
        "call_timeout": 0,
        "ping_timeout": 0,
        "rpc_pool_size": 0
    },
    "management_node": false,
    "auth_override": {
        "force_auth_provider": false,
        "auth_provider": {
            "name": "",
            "storage_engine": "",
            "meta": null
        },
        "force_session_provider": false,
        "session_provider": {
            "name": "",
            "storage_engine": "",
            "meta": null
        }
    },
    "enable_non_transactional_rate_limiter": false,
    "enable_sentinel_rate_limiter": false,
    "enable_redis_rolling_limiter": false,
    "drl_notification_frequency": 0,
    "drl_enable_sentinel_rate_limiter": false,
    "drl_threshold": 0,
    "enforce_org_data_age": false,
    "enforce_org_data_detail_logging": false,
    "enforce_org_quotas": false,
    "experimental_process_org_off_thread": false,
    "monitor": {
        "enable_trigger_monitors": false,
        "configuration": {
            "method": "",
            "target_path": "",
            "template_path": "",
            "header_map": null,
            "event_timeout": 0
        },
        "global_trigger_limit": 0,
        "monitor_user_keys": false,
        "monitor_org_keys": false
    },
    "max_idle_connections": 0,
    "max_idle_connections_per_host": 50,
    "max_conn_time": 90,
    "close_idle_connections": false,
    "close_connections": false,
    "enable_custom_domains": false,
    "allow_master_keys": false,
    "service_discovery": {
        "default_cache_timeout": 0
    },
    "proxy_ssl_insecure_skip_verify": false,
    "proxy_enable_http2": false,
    "proxy_ssl_min_version": 0,
    "proxy_ssl_ciphers": null,
    "proxy_default_timeout": 0,
    "proxy_ssl_disable_renegotiation": false,
    "proxy_close_connections": false,
    "uptime_tests": {
        "disable": false,
        "poller_group": "",
        "config": {
            "failure_trigger_sample_size": 0,
            "time_wait": 0,
            "checker_pool_size": 0,
            "enable_uptime_analytics": false
        }
    },
    "health_check": {
        "enable_health_checks": false,
        "health_check_value_timeouts": 0
    },
    "oauth_refresh_token_expire": 0,
    "oauth_token_expire": 0,
    "oauth_token_expired_retain_period": 0,
    "oauth_redirect_uri_separator": "",
    "oauth_error_status_code": 0,
    "enable_key_logging": false,
    "ssl_force_common_name_check": false,
    "enable_analytics": false,
    "analytics_config": {
        "type": "",
        "ignored_ips": [],
        "enable_detailed_recording": false,
        "enable_geo_ip": false,
        "geo_ip_db_path": "",
        "normalise_urls": {
            "enabled": false,
            "normalise_uuids": false,
            "normalise_numbers": false,
            "custom_patterns": null
        },
        "pool_size": 0,
        "records_buffer_size": 0,
        "storage_expiration_time": 0
    },
    "dns_cache": {
        "enabled": false,
        "ttl": 3600,
        "multiple_ips_handle_strategy": "no_cache"
    },
    "disable_regexp_cache": false,
    "regexp_cache_expire": 0,
    "local_session_cache": {
        "disable_cached_session_state": false,
        "cached_session_timeout": 0,
        "cached_session_eviction": 0
    },
    "enable_separate_cache_store": false,
    "cache_storage": {
        "type": "",
        "host": "",
        "port": 0,
        "hosts": null,
        "addrs": null,
        "master_name": "",
        "username": "",
        "password": "",
        "database": 0,
        "optimisation_max_idle": 0,
        "optimisation_max_active": 0,
        "timeout": 0,
        "enable_cluster": false,
        "use_ssl": false,
        "ssl_insecure_skip_verify": false
    },
    "enable_bundle_downloader": false,
    "bundle_base_url": "",
    "bundle_insecure_skip_verify": false,
    "enable_jsvm": false,
    "jsvm_timeout": 0,
    "disable_virtual_path_blobs": false,
    "tyk_js_path": "",
    "middleware_path": "middleware",
    "coprocess_options": {
        "enable_coprocess": false,
        "coprocess_grpc_server": "",
        "grpc_recv_max_size": 0,
        "grpc_send_max_size": 0,
        "python_path_prefix": "",
        "python_version": ""
    },
    "ignore_endpoint_case": false,
    "log_level": "info",
    "health_check_endpoint_name": "",
    "tracing": {
        "name": "",
        "enabled": false,
        "options": null
    },
    "newrelic": {
        "app_name": "",
        "license_key": ""
    },
    "enable_http_profiler": false,
    "use_redis_log": false,
    "sentry_code": "",
    "sentry_log_level": "",
    "use_sentry": false,
    "use_syslog": false,
    "use_graylog": false,
    "use_logstash": false,
    "graylog_network_addr": "",
    "logstash_network_addr": "",
    "syslog_transport": "",
    "logstash_transport": "",
    "syslog_network_addr": "",
    "statsd_connection_string": "",
    "statsd_prefix": "",
    "event_handlers": {
        "events": null
    },
    "event_trigers_defunct": null,
    "event_triggers_defunct": null,
    "session_update_pool_size": 0,
    "session_update_buffer_size": 0,
    "suppress_default_org_store": false,
    "legacy_enable_allowance_countdown": false,
    "global_session_lifetime": 0,
    "force_global_session_lifetime": false,
    "hide_generator_header": false,
    "kv": {
        "consul": {
            "address": "",
            "scheme": "",
            "datacenter": "",
            "http_auth": {
                "username": "",
                "password": ""
            },
            "wait_time": 0,
            "token": "",
            "tls_config": {
                "address": "",
                "ca_file": "",
                "ca_path": "",
                "cert_file": "",
                "key_file": "",
                "insecure_skip_verify": false
            }
        },
        "vault": {
            "address": "",
            "agent_address": "",
            "max_retries": 0,
            "timeout": 0,
            "token": "",
            "kv_version": 0
        }
    },
    "secrets": null,
    "override_messages": null
}
```

Against this API;

```json
{
  "created_at": "2023-07-12T13:48:50Z",
  "api_model": {},
  "api_definition": {
    "api_id": "88509231e7214a5d794881ce1c20443a",
    "jwt_issued_at_validation_skew": 0,
    "upstream_certificates": {},
    "use_keyless": true,
    "enable_coprocess_auth": false,
    "base_identity_provided_by": "",
    "custom_middleware": {
      "pre": [],
      "post": [],
      "post_key_auth": [],
      "auth_check": {
        "name": "",
        "path": "",
        "require_session": false,
        "raw_body_only": false
      },
      "response": [],
      "driver": "",
      "id_extractor": {
        "extract_from": "",
        "extract_with": "",
        "extractor_config": {}
      }
    },
    "disable_quota": false,
    "custom_middleware_bundle": "",
    "cache_options": {
      "cache_timeout": 60,
      "enable_cache": true,
      "cache_all_safe_requests": false,
      "cache_response_codes": [],
      "enable_upstream_cache_control": false,
      "cache_control_ttl_header": "",
      "cache_by_headers": []
    },
    "enable_ip_blacklisting": false,
    "tag_headers": [],
    "jwt_scope_to_policy_mapping": {},
    "pinned_public_keys": {},
    "expire_analytics_after": 0,
    "domain": "",
    "openid_options": {
      "providers": [],
      "segregate_by_client": false
    },
    "jwt_policy_field_name": "",
    "enable_proxy_protocol": false,
    "jwt_default_policies": [],
    "active": true,
    "jwt_expires_at_validation_skew": 0,
    "config_data": {},
    "notifications": {
      "shared_secret": "",
      "oauth_on_keychange_url": ""
    },
    "jwt_client_base_field": "",
    "auth": {
      "use_param": false,
      "param_name": "",
      "use_cookie": false,
      "cookie_name": "",
      "auth_header_name": "Authorization",
      "use_certificate": false,
      "validate_signature": false,
      "signature": {
        "algorithm": "",
        "header": "",
        "use_param": false,
        "param_name": "",
        "secret": "",
        "allowed_clock_skew": 0,
        "error_code": 0,
        "error_message": ""
      }
    },
    "check_host_against_uptime_tests": false,
    "auth_provider": {
      "name": "",
      "storage_engine": "",
      "meta": {}
    },
    "blacklisted_ips": [],
    "graphql": {
      "schema": "",
      "enabled": false,
      "engine": {
        "field_configs": [],
        "data_sources": []
      },
      "type_field_configurations": [],
      "execution_mode": "proxyOnly",
      "proxy": {
        "auth_headers": {}
      },
      "subgraph": {
        "sdl": ""
      },
      "supergraph": {
        "subgraphs": [],
        "merged_sdl": "",
        "global_headers": {},
        "disable_query_batching": false
      },
      "version": "2",
      "playground": {
        "enabled": false,
        "path": ""
      }
    },
    "hmac_allowed_clock_skew": -1,
    "dont_set_quota_on_create": false,
    "uptime_tests": {
      "check_list": [],
      "config": {
        "expire_utime_after": 0,
        "service_discovery": {
          "use_discovery_service": false,
          "query_endpoint": "",
          "use_nested_query": false,
          "parent_data_path": "",
          "data_path": "",
          "cache_timeout": 60
        },
        "recheck_wait": 0
      }
    },
    "enable_jwt": false,
    "do_not_track": false,
    "name": "dom",
    "slug": "dom",
    "oauth_meta": {
      "allowed_access_types": [],
      "allowed_authorize_types": [],
      "auth_login_redirect": ""
    },
    "CORS": {
      "enable": false,
      "max_age": 24,
      "allow_credentials": false,
      "exposed_headers": [],
      "allowed_headers": [
        "Origin",
        "Accept",
        "Content-Type",
        "X-Requested-With",
        "Authorization"
      ],
      "options_passthrough": false,
      "debug": false,
      "allowed_origins": [
        "*"
      ],
      "allowed_methods": [
        "GET",
        "POST",
        "HEAD"
      ]
    },
    "event_handlers": {
      "events": {}
    },
    "proxy": {
      "target_url": "http://localhost:9999",
      "service_discovery": {
        "endpoint_returns_list": false,
        "cache_timeout": 0,
        "parent_data_path": "",
        "query_endpoint": "",
        "use_discovery_service": false,
        "_sd_show_port_path": false,
        "target_path": "",
        "use_target_list": false,
        "use_nested_query": false,
        "data_path": "",
        "port_data_path": ""
      },
      "check_host_against_uptime_tests": false,
      "transport": {
        "ssl_insecure_skip_verify": false,
        "ssl_min_version": 0,
        "proxy_url": "",
        "ssl_ciphers": []
      },
      "target_list": [],
      "preserve_host_header": false,
      "strip_listen_path": true,
      "enable_load_balancing": false,
      "listen_path": "/dom/",
      "disable_strip_slash": true
    },
    "client_certificates": [],
    "use_basic_auth": false,
    "version_data": {
      "not_versioned": true,
      "default_version": "",
      "versions": {
        "Default": {
          "name": "Default",
          "expires": "",
          "paths": {
            "ignored": [],
            "white_list": [],
            "black_list": []
          },
          "use_extended_paths": true,
          "extended_paths": {
            "ignored": [],
            "white_list": [],
            "black_list": [],
            "transform": [],
            "transform_response": [],
            "transform_jq": [],
            "transform_jq_response": [],
            "transform_headers": [],
            "transform_response_headers": [],
            "hard_timeouts": [],
            "circuit_breakers": [],
            "url_rewrites": [],
            "virtual": [],
            "size_limits": [],
            "method_transforms": [],
            "track_endpoints": [],
            "do_not_track_endpoints": [],
            "validate_json": [],
            "internal": []
          },
          "global_headers": {},
          "global_headers_remove": [],
          "global_response_headers": {},
          "global_response_headers_remove": [],
          "ignore_endpoint_case": false,
          "global_size_limit": 0,
          "override_target": ""
        }
      }
    },
    "jwt_scope_claim_name": "",
    "use_standard_auth": false,
    "session_lifetime": 0,
    "hmac_allowed_algorithms": [],
    "disable_rate_limit": true,
    "definition": {
      "location": "header",
      "key": "x-api-version",
      "strip_path": false
    },
    "use_oauth2": false,
    "jwt_source": "",
    "jwt_signing_method": "",
    "jwt_not_before_validation_skew": 0,
    "use_go_plugin_auth": false,
    "jwt_identity_base_field": "",
    "allowed_ips": [],
    "request_signing": {
      "is_enabled": false,
      "secret": "",
      "key_id": "",
      "algorithm": "",
      "header_list": [],
      "certificate_id": "",
      "signature_header": ""
    },
    "org_id": "64aeae6c0a8726000111921d",
    "enable_ip_whitelisting": false,
    "global_rate_limit": {
      "rate": 0,
      "per": 0
    },
    "protocol": "",
    "enable_context_vars": false,
    "tags": [],
    "basic_auth": {
      "disable_caching": false,
      "cache_ttl": 0,
      "extract_from_body": false,
      "body_user_regexp": "",
      "body_password_regexp": ""
    },
    "listen_port": 0,
    "session_provider": {
      "name": "",
      "storage_engine": "",
      "meta": {}
    },
    "auth_configs": {
      "authToken": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      },
      "basic": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      },
      "coprocess": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      },
      "hmac": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      },
      "jwt": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      },
      "oauth": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      },
      "oidc": {
        "use_param": false,
        "param_name": "",
        "use_cookie": false,
        "cookie_name": "",
        "auth_header_name": "Authorization",
        "use_certificate": false,
        "validate_signature": false,
        "signature": {
          "algorithm": "",
          "header": "",
          "use_param": false,
          "param_name": "",
          "secret": "",
          "allowed_clock_skew": 0,
          "error_code": 0,
          "error_message": ""
        }
      }
    },
    "strip_auth_data": false,
    "id": "64aeaf420a8726000111921f",
    "certificates": [],
    "enable_signature_checking": false,
    "use_openid": false,
    "internal": false,
    "jwt_skip_kid": false,
    "enable_batch_request_support": false,
    "enable_detailed_recording": false,
    "response_processors": [],
    "use_mutual_tls_auth": false
  },
  "hook_references": [],
  "is_site": false,
  "sort_by": 0,
  "user_group_owners": [],
  "user_owners": []
}
```
I sent a total of `15,352,894` requests over `100` concurrent connections for `3000 seconds`. At the end of the test I was left with `53` idle connections and not 50;

```shell
tyk     38409  dom   20u     IPv4 55894681      0t0      TCP localhost:41940->localhost:distinct (ESTABLISHED)
tyk     38409  dom  115u     IPv4 55899549      0t0      TCP localhost:43976->localhost:distinct (ESTABLISHED)
tyk     38409  dom  119u     IPv4 55897466      0t0      TCP localhost:41724->localhost:distinct (ESTABLISHED)
tyk     38409  dom  120u     IPv4 55894822      0t0      TCP localhost:43926->localhost:distinct (ESTABLISHED)
tyk     38409  dom  124u     IPv4 55894821      0t0      TCP localhost:43906->localhost:distinct (ESTABLISHED)
tyk     38409  dom  125u     IPv4 55899499      0t0      TCP localhost:43444->localhost:distinct (ESTABLISHED)
tyk     38409  dom  126u     IPv4 55894820      0t0      TCP localhost:43894->localhost:distinct (ESTABLISHED)
tyk     38409  dom  127u     IPv4 55894819      0t0      TCP localhost:43878->localhost:distinct (ESTABLISHED)
tyk     38409  dom  128u     IPv4 55898573      0t0      TCP localhost:44222->localhost:distinct (ESTABLISHED)
tyk     38409  dom  130u     IPv4 55894840      0t0      TCP localhost:44050->localhost:distinct (ESTABLISHED)
tyk     38409  dom  131u     IPv4 55894816      0t0      TCP localhost:43848->localhost:distinct (ESTABLISHED)
tyk     38409  dom  132u     IPv4 52291096      0t0      TCP localhost:37502->localhost:distinct (ESTABLISHED)
tyk     38409  dom  135u     IPv4 55894811      0t0      TCP localhost:43810->localhost:distinct (ESTABLISHED)
tyk     38409  dom  137u     IPv4 55900495      0t0      TCP localhost:43788->localhost:distinct (ESTABLISHED)
tyk     38409  dom  138u     IPv4 55900376      0t0      TCP localhost:42214->localhost:distinct (ESTABLISHED)
tyk     38409  dom  139u     IPv4 55903168      0t0      TCP localhost:44296->localhost:distinct (ESTABLISHED)
tyk     38409  dom  142u     IPv4 55897681      0t0      TCP localhost:44262->localhost:distinct (ESTABLISHED)
tyk     38409  dom  143u     IPv4 55897675      0t0      TCP localhost:44242->localhost:distinct (ESTABLISHED)
tyk     38409  dom  144u     IPv4 55894815      0t0      TCP localhost:43844->localhost:distinct (ESTABLISHED)
tyk     38409  dom  145u     IPv4 55894786      0t0      TCP localhost:43602->localhost:distinct (ESTABLISHED)
tyk     38409  dom  146u     IPv4 55897663      0t0      TCP localhost:44168->localhost:distinct (ESTABLISHED)
tyk     38409  dom  148u     IPv4 55902995      0t0      TCP localhost:44044->localhost:distinct (ESTABLISHED)
tyk     38409  dom  149u     IPv4 55899547      0t0      TCP localhost:43968->localhost:distinct (ESTABLISHED)
tyk     38409  dom  150u     IPv4 55894757      0t0      TCP localhost:43226->localhost:distinct (ESTABLISHED)
tyk     38409  dom  153u     IPv4 55897682      0t0      TCP localhost:44328->localhost:distinct (ESTABLISHED)
tyk     38409  dom  155u     IPv4 55894708      0t0      TCP localhost:42432->localhost:distinct (ESTABLISHED)
tyk     38409  dom  157u     IPv4 55903169      0t0      TCP localhost:44302->localhost:distinct (ESTABLISHED)
tyk     38409  dom  159u     IPv4 51322008      0t0      TCP localhost:48712->localhost:distinct (ESTABLISHED)
tyk     38409  dom  164u     IPv4 55902568      0t0      TCP localhost:43164->localhost:distinct (ESTABLISHED)
tyk     38409  dom  169u     IPv4 55902996      0t0      TCP localhost:44066->localhost:distinct (ESTABLISHED)
tyk     38409  dom  173u     IPv4 55899282      0t0      TCP localhost:40906->localhost:distinct (ESTABLISHED)
tyk     38409  dom  176u     IPv4 55894827      0t0      TCP localhost:43942->localhost:distinct (ESTABLISHED)
tyk     38409  dom  177u     IPv4 55899553      0t0      TCP localhost:44008->localhost:distinct (ESTABLISHED)
tyk     38409  dom  179u     IPv4 55899491      0t0      TCP localhost:43394->localhost:distinct (ESTABLISHED)
tyk     38409  dom  180u     IPv4 55897594      0t0      TCP localhost:43322->localhost:distinct (ESTABLISHED)
tyk     38409  dom  183u     IPv4 55894785      0t0      TCP localhost:43586->localhost:distinct (ESTABLISHED)
tyk     38409  dom  184u     IPv4 55899551      0t0      TCP localhost:43996->localhost:distinct (ESTABLISHED)
tyk     38409  dom  188u     IPv4 55896781      0t0      TCP localhost:41156->localhost:distinct (ESTABLISHED)
tyk     38409  dom  189u     IPv4 55894668      0t0      TCP localhost:41668->localhost:distinct (ESTABLISHED)
tyk     38409  dom  190u     IPv4 55899372      0t0      TCP localhost:41890->localhost:distinct (ESTABLISHED)
tyk     38409  dom  192u     IPv4 55899485      0t0      TCP localhost:43338->localhost:distinct (ESTABLISHED)
tyk     38409  dom  197u     IPv4 55894659      0t0      TCP localhost:41608->localhost:distinct (ESTABLISHED)
tyk     38409  dom  198u     IPv4 55901879      0t0      TCP localhost:43598->localhost:distinct (ESTABLISHED)
tyk     38409  dom  199u     IPv4 55893570      0t0      TCP localhost:42664->localhost:distinct (ESTABLISHED)
tyk     38409  dom  203u     IPv4 55900513      0t0      TCP localhost:44040->localhost:distinct (ESTABLISHED)
tyk     38409  dom  206u     IPv4 55897651      0t0      TCP localhost:44102->localhost:distinct (ESTABLISHED)
tyk     38409  dom  211u     IPv4 55898466      0t0      TCP localhost:43002->localhost:distinct (ESTABLISHED)
tyk     38409  dom  214u     IPv4 55894863      0t0      TCP localhost:44148->localhost:distinct (ESTABLISHED)
tyk     38409  dom  216u     IPv4 55902553      0t0      TCP localhost:42820->localhost:distinct (ESTABLISHED)
tyk     38409  dom  218u     IPv4 55893592      0t0      TCP localhost:42822->localhost:distinct (ESTABLISHED)
tyk     38409  dom  221u     IPv4 55394508      0t0      TCP localhost:42864->localhost:distinct (ESTABLISHED)
tyk     38409  dom  222u     IPv4 55899501      0t0      TCP localhost:43496->localhost:distinct (ESTABLISHED)
tyk     38409  dom  230u     IPv4 55897601      0t0      TCP localhost:43532->localhost:distinct (ESTABLISHED)
```
So there does appear to be still a slight leak. Can you confirm you get the same?

Reply:
I think there's a race condition where;
- a connection is queued in idle queue before `DisableKeepAlive=true` is set
- `oldTransport.CloseIdleConnections()` doesn't close the active connection as not idle yet
- results in leaking idle connection once response is read

If I set the idle connection timeout
```golang
IdleConnTimeout:       time.Duration(p.Gw.GetConfig().MaxConnTime) * time.Second,
```
Whole function now reads;
```golang
func (p *ReverseProxy) defaultTransport(dialerTimeout float64) *http.Transport {
	timeout := 30.0
	if dialerTimeout > 0 {
		log.Debug("Setting timeout for outbound request to: ", dialerTimeout)
		timeout = dialerTimeout
	}

	dialer := &net.Dialer{
		Timeout:   time.Duration(float64(timeout) * float64(time.Second)),
		KeepAlive: 30 * time.Second,
		DualStack: true,
	}
	dialContextFunc := dialer.DialContext
	if p.Gw.dnsCacheManager.IsCacheEnabled() {
		dialContextFunc = p.Gw.dnsCacheManager.WrapDialer(dialer)
	}

	return &http.Transport{
		DialContext:           dialContextFunc,
		IdleConnTimeout:       time.Duration(p.Gw.GetConfig().MaxConnTime) * time.Second,
		MaxIdleConns:          p.Gw.GetConfig().MaxIdleConns,
		MaxIdleConnsPerHost:   p.Gw.GetConfig().MaxIdleConnsPerHost, // default is 100
		ResponseHeaderTimeout: time.Duration(dialerTimeout) * time.Second,
		TLSHandshakeTimeout:   10 * time.Second,
	}
}
```
This fixes the issue for me as the leaky connecrions on oldTransport eventually timeout. Can you confirm it works for you @buger ?

Reply:
I think you are correct here. Description of this param says:

// IdleConnTimeout is the maximum amount of time an idle
// (keep-alive) connection will remain idle before closing
// itself.
// Zero means no limit.

**Zero means no limit.**

Also `DefaultTransport` have it set to 90s. So missing this value indeed an issue, even if outside this max_conn_time option. The only nuance that I do not want set it dependent to `MaxConnTime` value. Because if it set to 0, it will be no limit. And theoretically this leak can happen, if you trigger a lot of API reloads while getting the traffic, because API reload also re-creates transport. I'm going to set it to 90s, to the same value as Go has set to default HTTP client.

Reply:
Created v4.0.8-rc10